### ISSUE TITLE
[js] blit may throw "source too large" exception if length is changed

### ISSUE THREAD

**Issue by vroad** [HaxeFoundation/haxe#5474]
On js, specifying length in constructor is not allowed, but subclasses of `haxe.io.Bytes` could set this to a different value. OpenFL ByteArray sometimes stops working with this exception on js. Probably `blit()` should use `src.b.byteLength` (or `src.b.length`?) to get consistent behavior in this case.

**Comment by ncannasse** [comment]
I tried reading several times but still couldn't get it: could you post some actual code?

Also, your issue title is way too long :-)

**Comment by vroad** [comment]
I'll try creating some test code.

**Comment by vroad** [comment]
I tried creating test code from OpenFL ByteArray impl. This throws "Source too large" error only on js, but not on other targets such as cpp because `b.byteLength` and `length` no longer match in Bytes.

```hx
import haxe.io.Bytes;
import haxe.io.BytesData;

class ExtBytes extends Bytes {
    
    private var __length:Int;
    
    public function new(length:Int)
    {
        var bytes:Bytes = Bytes.alloc (length);
		#if js
        super(bytes.b.buffer);
		#else
		super(length, bytes.b);
		#end
        __length = length;
    }
    
    public function clear ():Void {
		
		length = 0;
		
	}
    
    public function __resize (size:Int) {
		
		if (size > __length) {
			
			var bytes = Bytes.alloc ((((size + 1) * 3) >> 1));
			var cacheLength = length;
			length = __length;
			bytes.blit (0, this, 0, __length);
			length = cacheLength;
			__setData (bytes);
			
		}
		
		if (length < size) {
			
			length = size;
			
		}
		
	}
    
    private inline function __setData (bytes:Bytes):Void {
		
		b = bytes.b;
		__length = bytes.length;
		
		#if js
		data = bytes.data;
		#end
		
	}
}

class Test {
    static function main() {
        var src:ExtBytes = new ExtBytes(50);
        src.clear();
        src.__resize(25);
        var dst:ExtBytes = new ExtBytes(20);
        dst.clear();
        dst.__resize(25);
        dst.blit(0, src, 0, 25);
    }
}
```
http://try.haxe.org/#F94C9

**Comment by vroad** [comment]
I fixed this by checking underlying Uint8Array's byteLength instead of `Bytes.length`. Should I make pull request instead? 

https://github.com/vroad/haxe/commit/4eddb4dc4c3376c395df688e756793afaea0dd0e

**Comment by tbilou** [comment]
We're having the same issue.
Tried the "fix" locally and it solved the problem.
@ncannasse how should we proceed here? pull request?

**Comment by ncannasse** [comment]
@tbilou yes a PR would work, thanks

**Comment by tbilou** [comment]
@vroad can you do the pull request?

**Comment by vroad** [comment]
@tbilou Created a PR. https://github.com/HaxeFoundation/haxe/pull/5499

**Comment by ncannasse** [comment]
Closed by 8c03ac18749a25e63547fd045bc5d5c03985b8de
